### ISSUE TITLE
Focal length

### ISSUE THREAD

**Issue by LeonardPatrick** [Liumouliu/Deep_blind_PnP#2]
Thank you so much for your code. 

It seems that the focal length '800' is not used in the code. I wonder how to project the 3d data to align the 2d data.

I use the validation data, and use project_2d = 3d * I_3; The result is as follows:

<img width="804" alt="Screen Shot 2021-01-12 at 5 24 11 PM" src="https://user-images.githubusercontent.com/22613141/104295151-0c9be580-54fb-11eb-8566-a311064b3886.png">

How can I make the projection align? Thank you so much.

**Comment by Liumouliu** [comment]
Hi, Thank you for your interest.

I have already normalized all 2D points to rays using the intrinsic matrix.

You just need to set K = np.float32(np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]])) to project 3D points.

Please refer to line 203-213 in the file main_test.py.

Best,
liu

**Comment by LeonardPatrick** [comment]
Hi, thank you for your nice reply. 

I use the 3d_data(1:2)/3d_data(3) as the 3d_data's projection. That has exactly the same function as your intrinsic matrix K, i.e. np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]).

![image](https://user-images.githubusercontent.com/22613141/104458393-4990d680-55e6-11eb-9c5a-fef8dae2b7d1.png)

Thank you so much!

Best regards,
Jin

**Comment by LeonardPatrick** [comment]
BTW, I want to ask whether the code can be adapted into the scenario, where the 3d data has no explicit 1-to-1 correspondences with 2d data. Now that I just test your model40 dataset, the simulated 2d point has a connection with a specific 3d point.

In my case, they are from different modalities. So 2d data just aligns with the 3d projection visually, but the point numbers are different and they don't have explicit correspondences. Do you think whether the code can be adapted in such a case? Thank you in advance.

Best regards,
Jin

**Comment by Liumouliu** [comment]
Yes, as I have written in the Q/A section, this method can be used in your case, which is a partial-to-partial matching problem. I have done experiments to validate it in a localization application, taking a varying number of 3D and 2D points as input.

Some hints you may find useful:

1. Always using the intrinsic matrix K to normalize 2D points. 

2. If the coordinates of your 3D points vary in a large 3D domain (e.g., in an autonomous driving scene), subtract each 3D point with the mean 3D coordinate. Note, they are only used as inputs to the network to find 2D-3D matches, remember to transform coordinates back when solving the pose using PnP-RANSAC.

3. You may want to train the network using your dataset. In that case, prepare the training dataset in the same format as mine. For inlier 2D-3D matches involved in training, use the ground-truth pose to project 3D points to the image, and use a threshold with Nearest Neighbour search to determine whether a 2D-3D match is an inlier. Use the inlier 2D-3D matches for training.

Best,
liu
